### PR TITLE
[Yul] Consider return variables to be zero initially.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libyul/yulOptimizerTests/fullInliner/multi_fun_callback.yul [hunk 1]
```diff
@@ -41,20 +41,12 @@
 //         h_t := 2
 //         mstore(7, h_t)
 //         let g_x_1 := 10
-//         let g_f_x_8 := 1
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullInliner/multi_fun_callback.yul]: Ok, this is weird - will have to investigate.

File: test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul [hunk 2]
```diff
@@ -463,12 +463,13 @@
 //         let _2 := 0
 //         let _485 := mload(_2)
 //         let abi_encode_pos := _1
+//         let abi_encode_end_67_610
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul:466]: Oh yeah - opposite of what I wanted to achieve :)

File: libyul/optimiser/SSAValueTracker.cpp [hunk 3]
```diff
@@ -33,12 +33,18 @@ void SSAValueTracker::operator()(Assignment const& _assignment)
 		m_values.erase(var.name);
 }
 
+void SSAValueTracker::operator()(FunctionDefinition const& _funDef)
+{
+	for (auto const& var: _funDef.returnVariables)
+		setValue(var.name, nullptr);
+	ASTWalker::operator()(_funDef);
```
Comment (chriseth) [inline libyul/optimiser/SSAValueTracker.cpp:40]: This line was missing in the first version.

File: libyul/optimiser/SSAValueTracker.cpp [hunk 4]
```diff
@@ -33,12 +33,18 @@ void SSAValueTracker::operator()(Assignment const& _assignment)
 		m_values.erase(var.name);
 }
 
+void SSAValueTracker::operator()(FunctionDefinition const& _funDef)
+{
+	for (auto const& var: _funDef.returnVariables)
+		setValue(var.name, nullptr);
```
Comment (leonardoalt) [inline libyul/optimiser/SSAValueTracker.cpp:39]: Didn't the other PR set to `zero` instead of `nullptr`?

Comment (chriseth) [reply libyul/optimiser/SSAValueTracker.cpp:39]: Take a look at the implementation of `setValue` - it uses a special zero expression if the second argument is nullptr.

File: test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul [hunk 5]
```diff
@@ -562,16 +561,15 @@
...
-//             src_16 := add(src_16, array_allo__558)
+//             dst_15 := add(dst_15, array_allo__559)
+//             src_16 := add(src_16, array_allo__559)
 //         }
 //     }
 //     function abi_decode_t_array$_t_uint256_$dyn_memory_ptr(offset_27, end_28) -> array_29
 //     {
 //         if iszero(slt(add(offset_27, 0x1f), end_28))
 //         {
-//             let _88 := 0
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul:573]: Ah, but these are at least nice :)

**Review by ekpyron — APPROVED** [review]
Looks good!

**Comment by chriseth** [comment]
I'm merging. This PR should have zero impact on any gas test, because the yul optimizer is disabled.

**Pull request merged**